homeassistant/components/workday/binary_sensor.py

```diff
@@ -103,7 +103,20 @@ def setup_platform(hass, config, add_entities, discovery_info=None):
     # Remove holidays
     try:
         for date in remove_holidays:
-            obj_holidays.pop(date)
+            # is this formatted as a date?
+            if dt.parse_date(date):
+                # remove holiday by date
+                removed = obj_holidays.pop(date)
+                _LOGGER.debug("Removed %s", date)
+            else:
+                try:
+                    # remove holiday by name
+                    _LOGGER.debug("Treating '%s' as named holiday", date)
+                    removed = obj_holidays.pop_named(date)
+                    for holiday in removed:
+                        _LOGGER.debug("Removed %s by name '%s'", holiday, date)
+                except KeyError as unmatched:
+                    _LOGGER.debug("No holiday found matching %s", unmatched)
```

Comment:
Should this have a higher logging level? I assume we want to inform the end-user of this configuration error?

Reply:
Thanks @frenck. 

I toyed with this in my head a bit.  The counterexample I imagined had a holiday that may not exist every year, but the user wants to remove it any time it does.

Perhaps I was overthinking it—maybe log a warning and let the user choose to lower the log level for the integration?

Reply:
I think using a warning is the least we can do. It's a pity that holidays only returns holidays for the current year I think

Reply:
> It's a pity that holidays only returns holidays for the current year I think

Oh wait we already control that, I wonder if it's worth pulling the holidays for the next 5 years or so?

https://github.com/home-assistant/core/blob/dea5b95655d1e99d558d948a8bc20a32f0994a20/homeassistant/components/workday/binary_sensor.py#L81-L82

Reply:
I do think we want to avoid the case where the user suddenly receives a configuration error on January 1.

Would we want to hardcode a number of years into the integration?  We'd need to include past years for some cases as well (jubilees and royal weddings come to mind).

Reply:
Well not sure what the solution is. But if we can't tell the configuration is correct or wrong, the user will be in flux if it will ever work. I don't think that that is an acceptable situation.

Reply:
I agree we cannot have a situation where a valid configuration becomes invalid on 1st January. Hackily L82 could become `obj_holidays = getattr(holidays, country)(years=range(year-10, year+10))` but that's storing 20 times the data we had before and obviously may break in the extreme case of `offset > 3650` (or if someone doesn't update/restart for 10 years!).

Reply:
On the other hand, if a removed holiday goes by a different name in a new year, the user _would_ want to know it will not be removed.  As much as it would be a nuisance for the valid configuration to suddenly become invalid, it seems better than the behavior that could result from hiding the issue.  

There doesn't appear to be a perfect solution for every case here, but I think it'll be best to log a warning.  As long as the integration continues to process configuration and otherwise works normally I don't think this will cause too much grief—users can always filter out the message if so inclined.

Reply:
I'm really undecided on this to be honest. I really like the feature but I think it's hard to get it versatile. I see two closely related issues: we can't reliably ensure that a configuration that doesn't produce warnings one day won't the next maybe this is not that big an issue. The second one is `holidays` is rather picky about casing and punctuation as best I can tell. For example: it will require "St. Andrew's Day" (which is actually wrong by British style guides...) "St Andrews Day" or "st. andrew's day" won't work. We can't reliably user `title` to fix the latter as other cases would fail (e.g. "Platinum Jubilee of Elizabeth II"), I guess just point the user towards the source to check the format? Checking the log after reload/restart just to confirm the format is correct is not ideal. This at least can be fixed by loading the holiday list a config validation and checking at that point.

Reply:
Right, I did some more testing and discovered that removing an invalid holiday by date (say today) was already throwing up a `KeyError` and so this PR is already actually in a better place than the integration already was. I think the best way forward be to move the new `try:` to include the entire `if` statement for popping holidays (rather than just the `else` case) and then go ahead with this and try and make the integration more versatile down the line.

Reply:
I'm not sure how we'd feel about introducing an additional dependency and further complexity to this integration, but maybe we could resolve this with a fuzzy string match (i.e. [fuzzywuzzy](https://github.com/seatgeek/fuzzywuzzy)).  In the case of a non-existent holiday being removed, pull out the "closest" match (over a certain threshold) and display it in the warning message:

`No holiday found matching "St Andrews Day", did you mean "St. Andrew's Day"?`

If we didn't want the dependency we could implement a "quick and dirty" function to find the "longest match" from the holiday list, though this may return some comically unhelpful suggestions:

`No holiday found matching "Festivus", did you mean "Första Maj"?`

Simpler still, (though far less graceful) we could just include the whole list of known holidays:

`No holiday found matching 'Festivus', must match one of New Year's Day (Observed), Martin Luther King Jr. Day, Washington's Birthday, Memorial Day, Independence Day, Independence Day (Observed), Labor Day, Columbus Day, Veterans Day, Thanksgiving.`

Either of these could make its way into config validation as well.

Reply:
I think this can be worked on down the line, as I mentioned in docs I think config flow with an options flow to pick holidays to remove seems like the best option to me.